pkg/registry/core/service/storage/rest.go

```diff
@@ -674,19 +680,32 @@ func initClusterIP(service *api.Service, allocator ipallocator.Interface) (bool,
 			// not really an internal error.
 			return false, errors.NewInternalError(fmt.Errorf("failed to allocate a serviceIP: %v", err))
 		}
+		allocatedIP = ip
 		service.Spec.ClusterIP = ip.String()
-		return true, nil
 	case service.Spec.ClusterIP != api.ClusterIPNone && service.Spec.ClusterIP != "":
 		// Try to respect the requested IP.
-		if err := allocator.Allocate(net.ParseIP(service.Spec.ClusterIP)); err != nil {
+		ip := net.ParseIP(service.Spec.ClusterIP)
+		if err := allocator.Allocate(ip); err != nil {
 			// TODO: when validation becomes versioned, this gets more complicated.
 			el := field.ErrorList{field.Invalid(field.NewPath("spec", "clusterIP"), service.Spec.ClusterIP, err.Error())}
 			return false, errors.NewInvalid(api.Kind("Service"), service.Name, el)
 		}
-		return true, nil
+		allocatedIP = ip
+	}
+
+	// assuming the object was valid prior to setting, always force the IPFamily
+	// to match the allocated IP at this point
+	if allocatedIP != nil {
+		if utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) {
```

Comment:
```suggestion
		if service.Spec.IPFamily != nil || utilfeature.DefaultFeatureGate.Enabled(features.IPv6DualStack) {
```

Should this also force the ipFamily field value if there's an existing non-nil value in the field? in a skewed cluster at 1.18 (gate off) and 1.19 (gate on), this allocation could happen in an update of an existing service from type=ExternalName to type=ClusterIP, and could be allocated an IP at odds with the IPFamily field value assigned by the 1.19 server.

Reply:
Oh wow that's evil.

Reply:
No, because gate off clears the value from storage.  So we would have a nil value here and end up writing a nil value to storage, which the 1.19 server would inherit from the previously allocated IP.

Reply:
The server processing an update will either have nil or a defaulted value by this point - we should probably programmer error guard against an empty family reaching this point if the gate is on.